File: governance/index/qa/validators/security-validator.ts
/**
 * Security Validator
 *
 * Detects security vulnerabilities in real-time:
 * - Hardcoded credentials
 * - SQL injection
 * - XSS vulnerabilities
 * - Path traversal
 * - Secrets exposure
 *
 * Avg latency: <30ms
 */

export class SecurityValidator {
  private patterns: SecurityPattern[];

  constructor() {
    this.patterns = this.loadSecurityPatterns();
  }

  async validate(context: ValidationContext): Promise<ValidationResult> {
    const { data } = context;
    const violations: string[] = [];
    const suggestions: string[] = [];
    let maxSeverity: Severity = 'low';

    // Convert data to string for pattern matching
    const dataString = this.stringify(data);

    // Run all security checks
    for (const pattern of this.patterns) {
      const matches = this.checkPattern(dataString, pattern);

      if (matches.length > 0) {
        violations.push(...matches.map(m => `${pattern.name}: ${m}`));
        suggestions.push(pattern.fix);

        if (this.isMoreSevere(pattern.severity, maxSeverity)) {
          maxSeverity = pattern.severity;
        }
      }
    }

    return {
      validatorName: 'security',
      passed: violations.length === 0,
      violations,
      severity: maxSeverity,
      suggestions: [...new Set(suggestions)] // Remove duplicates
    };
  }

  private loadSecurityPatterns(): SecurityPattern[] {
    return [
      // Hardcoded credentials
      {
        name: 'Hardcoded Password',
        pattern: /password\s*[=:]\s*['"][^'"]+['"]/gi,
        severity: 'critical',
        fix: 'Use environment variables or secure vault (e.g., process.env.PASSWORD)'
      },
      {
        name: 'Hardcoded API Key',
        pattern: /(?:api[_-]?key|apikey|access[_-]?token)\s*[=:]\s*['"][^'"]+['"]/gi,
        severity: 'critical',
        fix: 'Store API keys in environment variables or secrets manager'
      },
      {
        name: 'Bearer Token in Code',
        pattern: /bearer\s+[a-zA-Z0-9_-]{20,}/gi,
        severity: 'critical',
        fix: 'Never hardcode tokens; use secure credential storage'
      },

      // SQL Injection
      {
        name: 'SQL Injection Risk',
        pattern: /(?:SELECT|INSERT|UPDATE|DELETE).*(?:\+|concat)/gi,
        severity: 'critical',
        fix: 'Use parameterized queries or ORM instead of string concatenation'
      },
      {
        name: 'Raw SQL Query',
        pattern: /\.query\([^)]*\+[^)]*\)/gi,
        severity: 'high',
        fix: 'Use prepared statements: .query("SELECT * FROM users WHERE id = ?", [id])'
      },

      // XSS Vulnerabilities
      {
        name: 'Unsafe HTML Rendering',
        pattern: /innerHTML\s*=|dangerouslySetInnerHTML/gi,
        severity: 'high',
        fix: 'Sanitize user input or use safe rendering methods'
      },
      {
        name: 'eval() Usage',
        pattern: /\beval\s*\(/gi,
        severity: 'critical',
        fix: 'Avoid eval(); use JSON.parse() or alternative safe methods'
      },

      // Path Traversal
      {
        name: 'Path Traversal Risk',
        pattern: /\.\.[/\\]/g,
        severity: 'high',
        fix: 'Validate and sanitize file paths; use path.normalize()'
      },

      // Secrets Exposure
      {
        name: 'Private Key in Code',
        pattern: /-----BEGIN (?:RSA )?PRIVATE KEY-----/g,
        severity: 'critical',
        fix: 'Never commit private keys; use secure key management'
      },
      {
        name: 'AWS Access Key',
        pattern: /AKIA[0-9A-Z]{16}/g,
        severity: 'critical',
        fix: 'Revoke exposed key immediately and use IAM roles'
      },

      // Insecure Randomness
      {
        name: 'Weak Random Generation',
        pattern: /Math\.random\(\)/gi,
        severity: 'medium',
        fix: 'Use crypto.randomBytes() for security-sensitive randomness'
      },

      // Command Injection
      {
        name: 'Command Injection Risk',
        pattern: /exec\(|spawn\([^)]*\+/gi,
        severity: 'critical',
        fix: 'Avoid shell execution with user input; use safe alternatives'
      },

      // Debugging Code
      {
        name: 'Debug Code in Production',
        pattern: /console\.(log|debug|trace)/gi,
        severity: 'low',
        fix: 'Remove debug logs or use proper logger with levels'
      }
    ];
  }

  private checkPattern(data: string, pattern: SecurityPattern): string[] {
    const matches: string[] = [];
    const regex = new RegExp(pattern.pattern);
    let match;

    while ((match = regex.exec(data)) !== null) {
      matches.push(match[0]);
    }

    return matches;
  }

  private stringify(data: unknown): string {
    if (typeof data === 'string') return data;
    if (typeof data === 'object') return JSON.stringify(data, null, 2);
    return String(data);
  }

  private isMoreSevere(a: Severity, b: Severity): boolean {
    const levels = { low: 0, medium: 1, high: 2, critical: 3 };
    return levels[a] > levels[b];
  }
}

interface SecurityPattern {
  name: string;
  pattern: RegExp;
  severity: Severity;
  fix: string;
}

type Severity = 'low' | 'medium' | 'high' | 'critical';

interface ValidationContext {
  eventId: string;
  data: unknown;
  metadata?: Record<string, unknown>;
}

interface ValidationResult {
  validatorName: string;
  passed: boolean;
  violations: string[];
  severity?: Severity;
  suggestions?: string[];
}

File: workspace/src/core/validators/security_validator.py
#!/usr/bin/env python3
"""Security Validator - Security vulnerability scanning"""
import re
import logging
from typing import Dict, Any, List, Tuple

logger = logging.getLogger(__name__)

class SecurityValidator:
    """Validates security aspects"""
    
    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.patterns = self._load_security_patterns()
    
    def validate(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Validate security"""
        code = data.get("code", "")
        errors = []
        warnings = []
        
        # Check for hardcoded secrets
        secret_issues = self._check_hardcoded_secrets(code)
        errors.extend(secret_issues)
        
        # Check for SQL injection
        sql_issues = self._check_sql_injection(code)
        errors.extend(sql_issues)
        
        # Check for XSS vulnerabilities
        xss_issues = self._check_xss(code)
        warnings.extend(xss_issues)
        
        return {
            "layer": "security",
            "is_valid": len(errors) == 0,
            "errors": errors,
            "warnings": warnings
        }
    
    def _load_security_patterns(self) -> List[Tuple[str, str, str]]:
        """Load security patterns"""
        return [
            ("hardcoded_secrets", r"(password|api_key|secret)\s*=\s*['\"]\w+", "critical"),
            ("sql_injection", r"execute\(.*\+.*\)", "critical"),
            ("xss", r"innerHTML\s*=\s*", "high"),
        ]
    
    def _check_hardcoded_secrets(self, code: str) -> List[str]:
        """Check for hardcoded secrets"""
        errors = []
        pattern = r"(password|api_key|secret|token)\s*=\s*['\"][\w-]+"
        matches = re.finditer(pattern, code, re.IGNORECASE)
        for match in matches:
            errors.append(f"Hardcoded secret found: {match.group(0)}")
        return errors
    
    def _check_sql_injection(self, code: str) -> List[str]:
        """Check for SQL injection patterns"""
        errors = []
        pattern = r"execute\([^)]*\+[^)]*\)"
        matches = re.finditer(pattern, code)
        for match in matches:
            errors.append(f"Potential SQL injection: {match.group(0)}")
        return errors
    
    def _check_xss(self, code: str) -> List[str]:
        """Check for XSS vulnerabilities"""
        warnings = []
        pattern = r"innerHTML\s*=\s*"
        matches = re.finditer(pattern, code)
        for match in matches:
            warnings.append(f"Potential XSS vulnerability: {match.group(0)}")
        return warnings
